### PR TITLE
Migration

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
First step of code consolidation is complete. Structural components of ctp have been placed in xtp includes classes:
Topology
Molecule
Segment
SegmentType
Fragment
Atom
QMNDList
QMPair
APolarSite
PolarSite

The migration branch will be left open for future migration of ctp classes as are deemed necessary. I believe @gtirimbo will be working on this.

### CODE REVIEW

**Comment by JoshuaSBrown** [comment]
Nothing in xtp is as of yet dependent on these classes, xtp even with this merge will still be dependent on the ctp structural classes that are listed above.  This does not include all the ctp classes that will be moved. I want to merge now so we are proceeding in baby steps as opposed to large overwhelming steps. :) 

The next step in the consolidation will deal with appropriately linking xtp with these structural classes thus breaking the ctp dependency and ensuring the unit tests still behave correctly. The linking will occur on a separate branch. Once linking has been established code refactoring will follow.

File: include/votca/xtp/qmnblist.h [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+public:
+    
+    /**
+     * \brief Container for records of type Donor-Bridge1-Bridge2-...-Acceptor
+     * 
+     * Every SuperExchangeType record contains a pair donor, acceptor
+     * and a list of bridges (all specified by segment types, i.e. strings)
+     * 
+     */ 
+    class SuperExchangeType {
```
Comment (JensWehner) [inline include/votca/xtp/qmnblist.h]: remove SuperExchange and everything that belongs to it

File: include/votca/xtp/qmpair.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+class QMPair : public std::pair< Segment*, Segment* >
+{
+public:
+    
+    enum PairType 
+    { 
+        Hopping,
+        SuperExchange,
+        SuperExchangeAndHopping,
```
Comment (JensWehner) [inline include/votca/xtp/qmpair.h]: should remove SuperExchange here as well

File: include/votca/xtp/qmpair.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+   void     setIsPathCarrier(bool yesno, int carrier);
+   bool     isPathCarrier(int carrier);
+
+   void     setLambdaO(double lO, int carrier);
+   double   getLambdaO(int carrier);
+   
+   double   getReorg12(int state) { return first->getU_nC_nN(state) + second->getU_cN_cC(state); } // 1->2
+   double   getReorg21(int state) { return first->getU_cN_cC(state) + second->getU_nC_nN(state); } // 2->1
+  
+   double   getReorg12_x(int state) { return first->getU_nX_nN(state) + second->getU_xN_xX(state); } // 1->2
```
Comment (JensWehner) [inline include/votca/xtp/qmpair.h]: get Reorg12 and getReorg12x should merge and we should put state into an enum

File: include/votca/xtp/qmpair.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                                       -first->getSiteEnergy(state); }
+
+   Segment* Seg1PbCopy() { return first; }
+   Segment* Seg2PbCopy();
+   Segment* Seg1() { return first; }
+   Segment* Seg2() { return second; }
+
+   bool     HasGhost() { return _hasGhost; }
+
+   // superexchange pairs have a list of bridging segments
```
Comment (JensWehner) [inline include/votca/xtp/qmpair.h]: remove superexchange

File: include/votca/xtp/segment.h [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+#include <map>
+#include <vector>
+
+#include <votca/tools/vec.h>
+
+namespace votca { namespace xtp {
+
+class Atom;
+class PolarSite;
```
Comment (JensWehner) [inline include/votca/xtp/segment.h]: remove PolarSite

File: include/votca/xtp/segment.h [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    void             calcPos();
+    void             TranslateBy(const tools::vec &shift);
+    
+    void            calcApproxSize();
+    double          getApproxSize(){return _approxsize;}
+
+    void             setHasState(bool yesno, int state);
+    bool             hasState(int state);
+
+    double           getOcc(int e_h_s_t);
```
Comment (JensWehner) [inline include/votca/xtp/segment.h]: use Enum instead of state

File: include/votca/xtp/segment.h [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    inline void      setType(SegmentType *type) { _typ = type; }
+    SegmentType     *getType() { return _typ; }
+
+    void             AddFragment( Fragment* fragment );
+    void             AddAtom( Atom* atom );
+    void             AddPolarSite(PolarSite *pole);
+    void             AddAPolarSite(APolarSite *pole);
+    std::vector< Fragment* > &Fragments() { return _fragments; }
+    std::vector < Atom* >    &Atoms() { return _atoms; }
+    std::vector<PolarSite*>  &PolarSites() { return _polarSites; }
```
Comment (JensWehner) [inline include/votca/xtp/segment.h]: remove PolarSites

File: include/votca/xtp/segment.h [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    std::vector < Fragment* >    _fragments;
+    std::vector < Atom* >        _atoms;
+    std::vector < PolarSite* >   _polarSites;
+    std::vector < APolarSite* >  _apolarSites;
+
+    tools::vec         _CoM;
+    double   _approxsize;
+
+
+    double _U_cC_nN_e;   // from ::EInternal     input     DEFAULT 0
```
Comment (JensWehner) [inline include/votca/xtp/segment.h]: maybe put all of the organisation energies into a vector later on to make the whole thing more clean

File: include/votca/xtp/topology.h [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+namespace votca { namespace xtp {
+
+class Molecule;
+class Segment;
+class SegmentType;
+class Fragment;
+class Atom;
+class APolarSite;
+class PolarSite;
```
Comment (JensWehner) [inline include/votca/xtp/topology.h]: remove

File: include/votca/xtp/topology.h [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+#include <string>
+
+#include <votca/tools/property.h>
+
+#include <votca/csg/boundarycondition.h>
+#include <votca/csg/openbox.h>
+#include <votca/csg/orthorhombicbox.h>
+#include <votca/csg/triclinicbox.h>
+
+#include <votca/moo/jcalc.h>
```
Comment (JensWehner) [inline include/votca/xtp/topology.h]: why is this here, any idea?

File: src/libxtp/qmpair.cc [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+
+  double QMPair::getLambdaO(int state) {
+    double result;
+    if (state ==-1) result=_lambdaO_e;
+    else if (state ==+1) result=_lambdaO_h;
+    else if (state ==+2) result=_lambdaO_s;
+    else if (state ==+3) result=_lambdaO_t;
+    else throw std::runtime_error(" ERROR CODE whx__01l1o__");
```
Comment (JensWehner) [inline src/libxtp/qmpair.cc]: I do not know what these error codes mean. Maybe a better Error message. We should store lambda in a vector and then get the number via an enum

File: src/libxtp/segment.cc [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+    _U_xN_xX_s = dU;
+  } else if (state == +3) {
+    _U_xN_xX_t = dU;
+  } else {
+    throw runtime_error(
+        " ERROR CODE whe__00u11d__");  // blabla?? What do I do here?
+  }
+}
+
+const double &Segment::getU_xX_nN(int state) {
```
Comment (JensWehner) [inline src/libxtp/segment.cc]: weshould probably have a common accessor for all of these

File: src/libxtp/topology.cc [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+
+  }
+
+
+  // ++++++++ //
+  // Populate //
+  // ++++++++ //
+
+  Fragment *Topology::AddFragment(string fragment_name) {
```
Comment (JensWehner) [inline src/libxtp/topology.cc]: should we move from pointer to visitor,unique and shared pointers?

File: src/libxtp/topology.cc [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+    // Determine box type automatically in case boxtype == typeAuto
+    if(boxtype == CSG::BoundaryCondition::typeAuto) {
+      boxtype = AutoDetectBoxType(box);
+    }
+
+    if(_hasPb) {
+      if (votca::tools::globals::verbose) {
+        cout << "Removing periodic box. Creating new... " << endl;
+      }
+      delete _bc;
```
Comment (JensWehner) [inline src/libxtp/topology.cc]: I do not really like this.

**Comment by JoshuaSBrown** [comment]
Why did I do what? Add the comment about the commit history. Well I don't want all the responsibility for what is in those files. :)

**Comment by JoshuaSBrown** [comment]
I'll get to your changes sometime tomorrow.

**Comment by JoshuaSBrown** [comment]
Ok, this can be merged when it is ready, we can deal with the other issues later.
